language/move-lang/src/cfgir/liveness/mod.rs

```diff
@@ -0,0 +1,260 @@
+                (*lbl, init)
+            })
+            .collect();
+        Liveness { states }
+    }
+}
+
+impl TransferFunctions for Liveness {
+    type State = LivenessState;
+
+    fn execute(
+        &mut self,
+        state: &mut Self::State,
+        label: Label,
+        idx: usize,
+        cmd: &Command,
+    ) -> Errors {
+        state.next_command();
+        command(state, cmd);
+        // set current [label][command_idx] data with the new liveness data
+        let cur_label_states = self.states.get_mut(&label).unwrap();
+        cur_label_states[idx] = state.clone();
+        Errors::new()
+    }
+}
+
+impl AbstractInterpreter for Liveness {}
+
+pub fn refine(
+    _signature: &FunctionSignature,
+    locals: &UniqueMap<Var, SingleType>,
+    cfg: &mut BlockCFG,
+    infinite_loop_starts: BTreeSet<Label>,
+) {
+    let (final_invariants, per_command_states) = {
+        let mut reverse = ReverseBlockCFG::new(cfg, infinite_loop_starts);
+        analyze(&mut reverse)
+    };
+    let initial_states = build_initial_states(cfg, final_invariants);
+    release_dead_refs(locals, initial_states, per_command_states, cfg.blocks_mut());
+}
+
+//**************************************************************************************************
+// Analysis
+//**************************************************************************************************
+
+fn analyze(reverse: &mut ReverseBlockCFG) -> (FinalInvariants, PerCommandStates) {
+    let initial_state = LivenessState::initial();
+    let mut liveness = Liveness::new(reverse);
+    let final_invariants = liveness.analyze_function(reverse, initial_state).unwrap();
+    (final_invariants, liveness.states)
+}
+
+fn build_initial_states(cfg: &BlockCFG, final_invariants: FinalInvariants) -> InitialStates {
+    let mut states = BTreeMap::new();
+    for lbl in cfg.blocks().keys() {
+        let mut state = LivenessState::initial();
+        for predecessor in cfg.predecessors(*lbl) {
+            // get the final liveness results for each predecessor, and union them
+            if let Some(pre) = final_invariants.get(predecessor) {
+                state.extend(pre)
+            }
+        }
+        states.insert(*lbl, state);
+    }
+    states
+}
+
+fn command(state: &mut LivenessState, sp!(_, cmd_): &Command) {
+    use Command_ as C;
+    match cmd_ {
+        C::Assign(ls, e) => {
+            lvalues(state, ls);
+            exp(state, e);
+        }
+        C::Mutate(el, er) => {
+            exp(state, er);
+            exp(state, el)
+        }
+        C::Return(e) | C::Abort(e) | C::IgnoreAndPop { exp: e, .. } | C::JumpIf { cond: e, .. } => {
+            exp(state, e)
+        }
+
+        C::Jump(_) => (),
+    }
+}
+
+fn lvalues(state: &mut LivenessState, ls: &[LValue]) {
+    ls.iter().for_each(|l| lvalue(state, l))
+}
+
+fn lvalue(state: &mut LivenessState, sp!(_, l_): &LValue) {
```

Comment:
What is `sp!(_, l_)`?

Reply:
It's a fake tuple for locations 
```
sp(loc, x): Spanned<X> // construction
sp!(loc, x) // patern matching ==> Spanned { loc, value: x }
```

Reply:
I might just be too much of a Rust n00b, but I was surprised that this was valid syntax/don't find it much easier to read now that I know it what it does. Might be simpler to take a normal parameter and do explicit pattern matching in the body?

Reply:
It's just a pattern match, and pattern matching is rather common on arguments in Rust. 

I use this style throughout the project, so I'll keep it consistent for this PR